Data sources from https://archive.ics.uci.edu/ml/machine-learning-databases/covtype/
Excercise in producing a decision tree to predict forest cover of parcels of land based on the 50+ tributes.
Based on the case study in Advnaced Analytics using Spark
Using Almond Shell kernel for Jupyter notebook

In [1]:
import $ivy.`org.apache.spark::spark-sql:2.4.3`
import org.apache.spark.sql._

Downloaded https://repo1.maven.org/maven2/sh/almond/almond-spark_2.12/0.8.2/almond-spark_2.12-0.8.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.3/spark-sql_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-parent_2.12/2.4.3/spark-parent_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/apache/18/apache-18.pom
Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-column/1.10.1/parquet-column-1.10.1.pom
Downloaded https://repo1.maven.org/maven2/sh/almond/ammonite-spark_2.12/0.7.2/ammonite-spark_2.12-0.7.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-catalyst_2.12/2.4.3/spark-catalyst_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/2.4.3/spark-tags_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/xbean/xbean-asm6-shaded/4.8/xbean-asm6-shaded-4.8.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spar

Downloaded https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.7.9/jackson-core-2.7.9.pom
Downloaded https://repo1.maven.org/maven2/org/apache/avro/avro-mapred/1.8.2/avro-mapred-1.8.2.pom
Downloaded https://repo1.maven.org/maven2/net/razorvine/pyrolite/4.13/pyrolite-4.13.pom
Downloaded https://repo1.maven.org/maven2/org/apache/orc/orc-shims/1.5.5/orc-shims-1.5.5.pom
Downloaded https://repo1.maven.org/maven2/javax/activation/activation/1.1.1/activation-1.1.1.pom
Downloaded https://repo1.maven.org/maven2/net/sf/py4j/py4j/0.10.7/py4j-0.10.7.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-network-shuffle_2.12/2.4.3/spark-network-shuffle_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/ivy/ivy/2.4.0/ivy-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/avro/avro/1.8.2/avro-1.8.2.pom
Downloaded https://repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-scala_2.12/2.6.7.1/jackson-module-scala_2.12-2.6.7.

Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-format/2.4.0/parquet-format-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.pom
Downloaded https://repo1.maven.org/maven2/org/antlr/antlr4-runtime/4.7/antlr4-runtime-4.7.pom
Downloaded https://repo1.maven.org/maven2/org/slf4j/jul-to-slf4j/1.7.16/jul-to-slf4j-1.7.16.pom
Downloaded https://repo1.maven.org/maven2/javax/servlet/javax.servlet-api/3.1.0/javax.servlet-api-3.1.0.pom
Downloaded https://repo1.maven.org/maven2/com/twitter/chill_2.12/0.9.3/chill_2.12-0.9.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/arrow/arrow-format/0.10.0/arrow-format-0.10.0.pom
Downloaded https://repo1.maven.org/maven2/commons-lang/commons-lang/2.6/commons-lang-2.6.pom
Downloaded https://repo1.maven.org/maven2/org/slf4j/slf4j-log4j12/1.7.16/slf4j-log4j12-1.7.16.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/core/jersey-common/2.22.2/jersey-common-2.22.2.po

Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/project/2.22.2/project-2.22.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-parent/41/commons-parent-41.pom
Downloaded https://repo1.maven.org/maven2/com/carrotsearch/hppc-parent/0.7.2/hppc-parent-0.7.2.pom
Downloaded https://repo1.maven.org/maven2/org/codehaus/janino/janino-parent/3.0.9/janino-parent-3.0.9.pom
Downloaded https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-project/9.4.20.v20190813/jetty-project-9.4.20.v20190813.pom
Downloaded https://repo1.maven.org/maven2/org/slf4j/slf4j-parent/1.7.25/slf4j-parent-1.7.25.pom
Downloaded https://repo1.maven.org/maven2/io/airlift/airbase/78/airbase-78.pom
Downloaded https://repo1.maven.org/maven2/com/fasterxml/jackson/jackson-parent/2.6.1/jackson-parent-2.6.1.pom
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-parent/40/commons-parent-40.pom
Downloaded https://repo1.maven.org/maven2/org/slf4j/slf4j-parent/1.7.16/slf4j-parent-

Downloaded https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-annotations/2.6.7/jackson-annotations-2.6.7.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/2.6.5/hadoop-common-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/javax/annotation/javax.annotation-api/1.2/javax.annotation-api-1.2.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-api/2.4.0-b34/hk2-api-2.4.0-b34.pom
Downloaded https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-http/9.4.20.v20190813/jetty-http-9.4.20.v20190813.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/bundles/repackaged/jersey-guava/2.22.2/jersey-guava-2.22.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-compress/1.8.1/commons-compress-1.8.1.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-app/2.6.5/hadoop-mapreduce-client-app-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/tukaani/xz/1.5/xz-1.5.p

Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-common/2.6.5/hadoop-mapreduce-client-common-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/mortbay/jetty/jetty-util/6.1.26/jetty-util-6.1.26.pom
Downloaded https://repo1.maven.org/maven2/com/google/code/gson/gson/2.2.4/gson-2.2.4.pom
Downloaded https://repo1.maven.org/maven2/org/htrace/htrace-core/3.0.4/htrace-core-3.0.4.pom
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-ast_2.12/3.5.3/json4s-ast_2.12-3.5.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/curator/curator-client/2.6.0/curator-client-2.6.0.pom
Downloaded https://repo1.maven.org/maven2/commons-httpclient/commons-httpclient/3.1/commons-httpclient-3.1.pom
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-scalap_2.12/3.5.3/json4s-scalap_2.12-3.5.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-shuffle/2.6.5/hadoop-mapreduce-client-shuffle-2.6.5.pom
Downloaded htt

Downloaded https://repo1.maven.org/maven2/org/apache/apache/11/apache-11.pom
Downloaded https://repo1.maven.org/maven2/com/google/inject/guice/3.0/guice-3.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/directory/api/api-asn1-api/1.0.0-M20/api-asn1-api-1.0.0-M20.pom
Downloaded https://repo1.maven.org/maven2/org/apache/httpcomponents/httpcore/4.2.4/httpcore-4.2.4.pom
Downloaded https://repo1.maven.org/maven2/javax/xml/stream/stax-api/1.0-2/stax-api-1.0-2.pom
Downloaded https://repo1.maven.org/maven2/org/codehaus/jettison/jettison/1.1/jettison-1.1.pom
Downloaded https://repo1.maven.org/maven2/commons-beanutils/commons-beanutils/1.7.0/commons-beanutils-1.7.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/directory/server/apacheds-i18n/2.0.0-M15/apacheds-i18n-2.0.0-M15.pom
Downloaded https://repo1.maven.org/maven2/jline/jline/0.9.94/jline-0.9.94.pom
Downloaded https://repo1.maven.org/maven2/org/apache/directory/api/api-util/1.0.0-M20/api-util-1.0.0-M20.pom
Downloaded h

Downloaded https://repo1.maven.org/maven2/commons-httpclient/commons-httpclient/3.1/commons-httpclient-3.1.jar
Downloaded https://repo1.maven.org/maven2/org/xerial/snappy/snappy-java/1.1.7.3/snappy-java-1.1.7.3.jar
Downloaded https://repo1.maven.org/maven2/log4j/log4j/1.2.17/log4j-1.2.17.jar
Downloaded https://repo1.maven.org/maven2/commons-net/commons-net/3.1/commons-net-3.1.jar
Downloaded https://repo1.maven.org/maven2/commons-beanutils/commons-beanutils/1.7.0/commons-beanutils-1.7.0.jar
Downloaded https://repo1.maven.org/maven2/commons-digester/commons-digester/1.8/commons-digester-1.8.jar
Downloaded https://repo1.maven.org/maven2/commons-collections/commons-collections/3.2.2/commons-collections-3.2.2.jar
Downloaded https://repo1.maven.org/maven2/commons-lang/commons-lang/2.6/commons-lang-2.6.jar
Downloaded https://repo1.maven.org/maven2/org/slf4j/slf4j-log4j12/1.7.16/slf4j-log4j12-1.7.16.jar
Downloaded https://repo1.maven.org/maven2/com/google/code/gson/gson/2.2.4/gson-2.2.4.jar
Do

Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-server-common/2.6.5/hadoop-yarn-server-common-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-core/2.6.5/hadoop-mapreduce-client-core-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-common/2.6.5/hadoop-yarn-common-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-shuffle/2.6.5/hadoop-mapreduce-client-shuffle-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/jettison/jettison/1.1/jettison-1.1.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-common/2.6.5/hadoop-mapreduce-client-common-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/2.6.5/hadoop-common-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-client/2.6.5/hadoop-yarn-client-2.6.5.jar
Downloaded https://repo1.maven.org

Downloaded https://repo1.maven.org/maven2/com/github/luben/zstd-jni/1.3.2-2/zstd-jni-1.3.2-2.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-utils/2.4.0-b34/hk2-utils-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/external/aopalliance-repackaged/2.4.0-b34/aopalliance-repackaged-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/external/javax.inject/2.4.0-b34/javax.inject-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-locator/2.4.0-b34/hk2-locator-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/osgi-resource-locator/1.0.1/osgi-resource-locator-1.0.1.jar
Downloaded https://repo1.maven.org/maven2/org/javassist/javassist/3.18.1-GA/javassist-3.18.1-GA.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-hdfs/2.6.5/hadoop-hdfs-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-api/2.4.0-b34/hk2-api-2.4.0-b34.jar
Downloaded ht

Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-hadoop/1.10.1/parquet-hadoop-1.10.1.jar
Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-encoding/1.10.1/parquet-encoding-1.10.1.jar
Downloaded https://repo1.maven.org/maven2/org/apache/arrow/arrow-format/0.10.0/arrow-format-0.10.0.jar
Downloaded https://repo1.maven.org/maven2/com/vlkan/flatbuffers/1.2.0-3f79e055/flatbuffers-1.2.0-3f79e055.jar
Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-column/1.10.1/parquet-column-1.10.1.jar
Downloaded https://repo1.maven.org/maven2/org/apache/arrow/arrow-memory/0.10.0/arrow-memory-0.10.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-jackson/1.10.1/parquet-jackson-1.10.1.jar
Downloaded https://repo1.maven.org/maven2/joda-time/joda-time/2.9.9/joda-time-2.9.9.jar
Downloaded https://repo1.maven.org/maven2/org/apache/arrow/arrow-vector/0.10.0/arrow-vector-0.10.0.jar
Downloaded https://repo1.maven.org/maven2/org/eclip

Downloaded https://repo1.maven.org/maven2/commons-cli/commons-cli/1.2/commons-cli-1.2-sources.jar
Downloaded https://repo1.maven.org/maven2/xmlenc/xmlenc/0.52/xmlenc-0.52-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/xbean/xbean-asm6-shaded/4.8/xbean-asm6-shaded-4.8-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-io/commons-io/2.4/commons-io-2.4-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-net/commons-net/3.1/commons-net-3.1-sources.jar
Downloaded https://repo1.maven.org/maven2/com/google/guava/guava/16.0.1/guava-16.0.1-sources.jar
Downloaded https://repo1.maven.org/maven2/org/xerial/snappy/snappy-java/1.1.7.3/snappy-java-1.1.7.3-sources.jar
Downloaded https://repo1.maven.org/maven2/log4j/log4j/1.2.17/log4j-1.2.17-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-collections/commons-collections/3.2.2/commons-collections-3.2.2-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.3/spark-sq

Downloaded https://repo1.maven.org/maven2/xml-apis/xml-apis/1.3.04/xml-apis-1.3.04-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/jackson/jackson-jaxrs/1.9.13/jackson-jaxrs-1.9.13-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-api/2.6.5/hadoop-yarn-api-2.6.5-sources.jar
Downloaded https://repo1.maven.org/maven2/javax/xml/bind/jaxb-api/2.2.2/jaxb-api-2.2.2-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/jackson/jackson-xc/1.9.13/jackson-xc-1.9.13-sources.jar
Downloaded https://repo1.maven.org/maven2/javax/inject/javax.inject/1/javax.inject-1-sources.jar
Downloaded https://repo1.maven.org/maven2/aopalliance/aopalliance/1.0/aopalliance-1.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/fusesource/leveldbjni/leveldbjni-all/1.8/leveldbjni-all-1.8-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-client/2.6.5/hadoop-yarn-client-2.6.5-sources.jar
Downloaded https://repo1.ma

Downloaded https://repo1.maven.org/maven2/org/slf4j/jcl-over-slf4j/1.7.16/jcl-over-slf4j-1.7.16-sources.jar
Downloaded https://repo1.maven.org/maven2/javax/servlet/javax.servlet-api/3.1.0/javax.servlet-api-3.1.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/lz4/lz4-java/1.4.0/lz4-java-1.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.6.7.1/jackson-databind-2.6.7.1-sources.jar
Downloaded https://repo1.maven.org/maven2/com/github/luben/zstd-jni/1.3.2-2/zstd-jni-1.3.2-2-sources.jar
Downloaded https://repo1.maven.org/maven2/org/roaringbitmap/shims/0.7.45/shims-0.7.45-sources.jar
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-ast_2.12/3.5.3/json4s-ast_2.12-3.5.3-sources.jar
Downloaded https://repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6-sources.jar
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-core_2.12/3.5.3/json4s-core_2.12-3.5.3-sources.jar
Downloaded 

Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/bundles/repackaged/jersey-guava/2.22.2/jersey-guava-2.22.2-sources.jar
Downloaded https://repo1.maven.org/maven2/oro/oro/2.0.8/oro-2.0.8-sources.jar
Downloaded https://repo1.maven.org/maven2/net/razorvine/pyrolite/4.13/pyrolite-4.13-sources.jar
Downloaded https://repo1.maven.org/maven2/net/sf/py4j/py4j/0.10.7/py4j-0.10.7-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/janino/commons-compiler/3.0.9/commons-compiler-3.0.9-sources.jar
Downloaded https://repo1.maven.org/maven2/org/scala-lang/modules/scala-parser-combinators_2.12/1.1.0/scala-parser-combinators_2.12-1.1.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.8/scala-reflect-2.12.8-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/janino/janino/3.0.9/janino-3.0.9-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/orc/orc-shims/1.5.5/orc-shims-1.5.5-sources.jar
Downloaded https://re

Downloaded https://repo1.maven.org/maven2/xerces/xercesImpl/2.9.1/xercesImpl-2.9.1-sources.jar.sha1
Downloaded https://repo1.maven.org/maven2/xmlenc/xmlenc/0.52/xmlenc-0.52-sources.jar.sha1
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-lang3/3.5/commons-lang3-3.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.3/spark-sql_2.12-2.4.3-sources.jar


import $ivy.$                                  

import org.apache.spark.sql._

In [2]:
import org.apache.log4j.{Logger,Level}

import org.apache.log4j.{Logger,Level}

In [3]:
Logger.getLogger("org").setLevel(Level.ERROR)

Create the sql sessions 

In [4]:
val spark = {
            org.apache.spark.sql.NotebookSparkSession.builder()
            .master("local[*]")
            .getOrCreate()
            }

Loading spark-stubs


Downloaded https://repo1.maven.org/maven2/sh/almond/spark-stubs_24_2.12/0.7.2/spark-stubs_24_2.12-0.7.2.pom
Downloaded https://repo1.maven.org/maven2/sh/almond/spark-stubs_24_2.12/0.7.2/spark-stubs_24_2.12-0.7.2-sources.jar
Downloaded https://repo1.maven.org/maven2/sh/almond/spark-stubs_24_2.12/0.7.2/spark-stubs_24_2.12-0.7.2.jar


Getting spark JARs
Creating SparkSession


Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties


Spark UI

spark: SparkSession = org.apache.spark.sql.SparkSession@4f11b10a

In [5]:
val data = spark.read
            .option("header","false")
            .option("inferSchema","true")
            .csv("../covtype/")

csv at cmd4.sc:4

1 / 1

csv at cmd4.sc:4

4 / 4

data: DataFrame = [_c0: string, _c1: string ... 53 more fields]

In [6]:
data.count()

count at cmd5.sc:1

4 / 4

count at cmd5.sc:1

1 / 1

res5: Long = 581259L

In [7]:
data.head()

head at cmd6.sc:1

1 / 1

res6: Row = [2596,51,3,258,0,510,221,232,148,6279,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,5]

In [8]:
data.printSchema()

root
 |-- _c0: string (nullable = true)
 |-- _c1: string (nullable = true)
 |-- _c2: string (nullable = true)
 |-- _c3: integer (nullable = true)
 |-- _c4: integer (nullable = true)
 |-- _c5: integer (nullable = true)
 |-- _c6: integer (nullable = true)
 |-- _c7: integer (nullable = true)
 |-- _c8: integer (nullable = true)
 |-- _c9: integer (nullable = true)
 |-- _c10: integer (nullable = true)
 |-- _c11: integer (nullable = true)
 |-- _c12: integer (nullable = true)
 |-- _c13: integer (nullable = true)
 |-- _c14: integer (nullable = true)
 |-- _c15: integer (nullable = true)
 |-- _c16: integer (nullable = true)
 |-- _c17: integer (nullable = true)
 |-- _c18: integer (nullable = true)
 |-- _c19: integer (nullable = true)
 |-- _c20: integer (nullable = true)
 |-- _c21: integer (nullable = true)
 |-- _c22: integer (nullable = true)
 |-- _c23: integer (nullable = true)
 |-- _c24: integer (nullable = true)
 |-- _c25: integer (nullable = true)
 |-- _c26: integer (nullable = true)
 |-- _c27

Looks like the schema isn't what was expected. _c0..2 should be integers

In [9]:
import org.apache.spark.sql.types._

import org.apache.spark.sql.types._

Have to update the scheme. Only way seems to be to create new DF every time.

In [10]:
val data2=data.withColumn("_c0",data("_c0").cast(IntegerType))
val data3=data2.withColumn("_c1",data2("_c1").cast(IntegerType))
val data4=data3.withColumn("_c2",data3("_c2").cast(IntegerType))

data2: DataFrame = [_c0: int, _c1: string ... 53 more fields]
data3: DataFrame = [_c0: int, _c1: int ... 53 more fields]
data4: DataFrame = [_c0: int, _c1: int ... 53 more fields]

In [11]:
data4.printSchema()

root
 |-- _c0: integer (nullable = true)
 |-- _c1: integer (nullable = true)
 |-- _c2: integer (nullable = true)
 |-- _c3: integer (nullable = true)
 |-- _c4: integer (nullable = true)
 |-- _c5: integer (nullable = true)
 |-- _c6: integer (nullable = true)
 |-- _c7: integer (nullable = true)
 |-- _c8: integer (nullable = true)
 |-- _c9: integer (nullable = true)
 |-- _c10: integer (nullable = true)
 |-- _c11: integer (nullable = true)
 |-- _c12: integer (nullable = true)
 |-- _c13: integer (nullable = true)
 |-- _c14: integer (nullable = true)
 |-- _c15: integer (nullable = true)
 |-- _c16: integer (nullable = true)
 |-- _c17: integer (nullable = true)
 |-- _c18: integer (nullable = true)
 |-- _c19: integer (nullable = true)
 |-- _c20: integer (nullable = true)
 |-- _c21: integer (nullable = true)
 |-- _c22: integer (nullable = true)
 |-- _c23: integer (nullable = true)
 |-- _c24: integer (nullable = true)
 |-- _c25: integer (nullable = true)
 |-- _c26: integer (nullable = true)
 |-- _

Above approach worked by duplicating the dataframe everytime seem thereis potential for memory overflow. Try doing it when adding titles.

Create header titles for the columns. 

In [12]:
val colNames = Seq(
"Elevation", "Aspect", "Slope",
"Horizontal_Distance_To_Hydrology", "Vertical_Distance_To_Hydrology",
"Horizontal_Distance_To_Roadways",
"Hillshade_9am", "Hillshade_Noon", "Hillshade_3pm",
"Horizontal_Distance_To_Fire_Points"
) ++ (
(0 until 4).map(i => s"Wilderness_Area_$i")
) ++ (
(0 until 40).map(i => s"Soil_Type_$i")
) ++ Seq("Cover_Type")

colNames: Seq[String] = List(
  "Elevation",
  "Aspect",
  "Slope",
  "Horizontal_Distance_To_Hydrology",
  "Vertical_Distance_To_Hydrology",
  "Horizontal_Distance_To_Roadways",
  "Hillshade_9am",
  "Hillshade_Noon",
  "Hillshade_3pm",
  "Horizontal_Distance_To_Fire_Points",
  "Wilderness_Area_0",
  "Wilderness_Area_1",
  "Wilderness_Area_2",
  "Wilderness_Area_3",
  "Soil_Type_0",
  "Soil_Type_1",
  "Soil_Type_2",
  "Soil_Type_3",
  "Soil_Type_4",
  "Soil_Type_5",
  "Soil_Type_6",
  "Soil_Type_7",
  "Soil_Type_8",
  "Soil_Type_9",
  "Soil_Type_10",
  "Soil_Type_11",
  "Soil_Type_12",
  "Soil_Type_13",
  "Soil_Type_14",
  "Soil_Type_15",
  "Soil_Type_16",
  "Soil_Type_17",
  "Soil_Type_18",
  "Soil_Type_19",
  "Soil_Type_20",
  "Soil_Type_21",
  "Soil_Type_22",
  "Soil_Type_23",
...

In [13]:
colNames.getClass

res12: Class[T] = class scala.collection.immutable.$colon$colon

implicits contains the function $ which returns a column object. Needed to access the data type in order to cast it to integer. $("string"): column   ... pretty 'hacky' syntax.

In [14]:
import spark.implicits._

import spark.implicits._

In [15]:
val dataClean = data.toDF(colNames:_*).
    withColumn("Elevation",$"Elevation".cast(IntegerType))
    .withColumn("Aspect",$"Aspect".cast(IntegerType))
    .withColumn("Slope",$"Slope".cast(IntegerType))
    .withColumn("Cover_Type",$"Cover_Type".cast(DoubleType))


dataClean: DataFrame = [Elevation: int, Aspect: int ... 53 more fields]

In [16]:
dataClean.show() // Using the spark.sql API

show at cmd15.sc:1

1 / 1

+---------+------+-----+--------------------------------+------------------------------+-------------------------------+-------------+--------------+-------------+----------------------------------+-----------------+-----------------+-----------------+-----------------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+------------+----------+
|Elevation|Aspect|Slope|Horizontal_Distance_To_Hydrology|Vertical_Distance_To_Hydrology|Horizontal_Distance_To_Roadways|Hillshade_9am|Hillshade_Noon|Hillshade_3pm|Horizontal_Distance_To_Fire_Points|Wilderness

|     2617|    45|    9|                             240|                            56|                            666|          223|           221|          133|                              6244|                1|                0|                0|                0|          0|          0|          0|          0|          0|          0|          0|          0|          0|          0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           1|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|       5.0|
|     2612|    59|   10|                             247|                            11|                            636|          228|           219|          124|                              6230|          

|     2504|     0|    4|                              95|                             5|                            691|          214|           232|          156|                              5572|                1|                0|                0|                0|          0|          0|          0|          0|          0|          0|          0|          0|          0|          0|           0|           0|           0|           0|           0|           0|           0|           1|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|           0|       5.0|
|     2503|    38|    5|                              85|                            10|                            741|          220|           228|          144|                              5555|          

Randomly split the data into train and test using 10% for testing (will need to further split the training data if CV performed

In [17]:
val Array(trainData, testData) = dataClean.randomSplit(Array(0.9,0.1))
trainData.cache()
testData.cache()

trainData: Dataset[Row] = [Elevation: int, Aspect: int ... 53 more fields]
testData: Dataset[Row] = [Elevation: int, Aspect: int ... 53 more fields]
res16_1: Dataset[Row] = [Elevation: int, Aspect: int ... 53 more fields]
res16_2: Dataset[Row] = [Elevation: int, Aspect: int ... 53 more fields]

Vectorise the features. This is required by all SparkML alogorithms.

Took some messing around to find the right repo for ml lib. Eventually found a reference on the Mvnrepository site to find the mllib version that goes with scala 2.12.

In [18]:
import $ivy.`org.apache.spark::spark-mllib:2.4.0`

Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-mllib_2.12/2.4.0/spark-mllib_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-parent_2.12/2.4.0/spark-parent_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-streaming_2.12/2.4.0/spark-streaming_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/scalanlp/breeze_2.12/0.13.2/breeze_2.12-0.13.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-graphx_2.12/2.4.0/spark-graphx_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-mllib-local_2.12/2.4.0/spark-mllib-local_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/2.4.0/spark-tags_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.0/spark-sql_2.12-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-core_2.12/2.4.0/spark-core_2.12-2.4.0.pom
Downloaded http

Downloaded https://repo1.maven.org/maven2/com/thoughtworks/paranamer/paranamer-parent/2.7/paranamer-parent-2.7.pom
Downloaded https://repo1.maven.org/maven2/org/codehaus/codehaus-parent/4/codehaus-parent-4.pom
Downloaded https://repo1.maven.org/maven2/commons-beanutils/commons-beanutils-core/1.8.0/commons-beanutils-core-1.8.0.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-kvstore_2.12/2.4.0/spark-kvstore_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-network-shuffle_2.12/2.4.0/spark-network-shuffle_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-unsafe_2.12/2.4.0/spark-unsafe_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-launcher_2.12/2.4.0/spark-launcher_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/2.4.0/spark-tags_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmap/0.5.11/RoaringBitma

Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-network-common_2.12/2.4.0/spark-network-common_2.12-2.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-unsafe_2.12/2.4.0/spark-unsafe_2.12-2.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.0/spark-sql_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmap/0.5.11/RoaringBitmap-0.5.11-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-streaming_2.12/2.4.0/spark-streaming_2.12-2.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sketch_2.12/2.4.0/spark-sketch_2.12-2.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-mllib_2.12/2.4.0/spark-mllib_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-catalyst_2.12/2.4.0/spark-catalyst_2.12-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/orc/or

import $ivy.$                                    

In [19]:
import org.apache.spark.ml.feature.VectorAssembler

import org.apache.spark.ml.feature.VectorAssembler

In [20]:
val inputCols = trainData.columns.filter(_!="Cover_Type")
val assembler = new org.apache.spark.ml.feature.VectorAssembler()
                    .setInputCols(inputCols)
                    .setOutputCol("featureVector")

inputCols: Array[String] = Array(
  "Elevation",
  "Aspect",
  "Slope",
  "Horizontal_Distance_To_Hydrology",
  "Vertical_Distance_To_Hydrology",
  "Horizontal_Distance_To_Roadways",
  "Hillshade_9am",
  "Hillshade_Noon",
  "Hillshade_3pm",
  "Horizontal_Distance_To_Fire_Points",
  "Wilderness_Area_0",
  "Wilderness_Area_1",
  "Wilderness_Area_2",
  "Wilderness_Area_3",
  "Soil_Type_0",
  "Soil_Type_1",
  "Soil_Type_2",
  "Soil_Type_3",
  "Soil_Type_4",
  "Soil_Type_5",
  "Soil_Type_6",
  "Soil_Type_7",
  "Soil_Type_8",
  "Soil_Type_9",
  "Soil_Type_10",
  "Soil_Type_11",
  "Soil_Type_12",
  "Soil_Type_13",
  "Soil_Type_14",
  "Soil_Type_15",
  "Soil_Type_16",
  "Soil_Type_17",
  "Soil_Type_18",
  "Soil_Type_19",
  "Soil_Type_20",
  "Soil_Type_21",
  "Soil_Type_22",
  "Soil_Type_23",
...
assembler: VectorAssembler = vecAssembler_d146f862bc99

Since spark 2.4 vector assembler extends HadHandleInvalid which requires you to specify how nulls will be handled. It take a single parameter ("skip") to skip when encountered. If not used then the dataset must be cleaned up and the missing data imputed. 

See stackoverflow
https://stackoverflow.com/questions/41362295/sparkexception-values-to-assemble-cannot-be-null


TODO: Investigate the source of nulls

In [21]:
val assembledTrainData = assembler.setHandleInvalid("skip").transform(trainData)

assembledTrainData: DataFrame = [Elevation: int, Aspect: int ... 54 more fields]

In [22]:
assembledTrainData.select("featureVector").show(truncate=false)

show at cmd21.sc:1

1 / 1

+-----------------------------------------------------------------------------------------------------+
|featureVector                                                                                        |
+-----------------------------------------------------------------------------------------------------+
|(54,[0,1,2,3,4,5,6,7,8,9,13,15],[1863.0,37.0,17.0,120.0,18.0,90.0,217.0,202.0,115.0,769.0,1.0,1.0])  |
|(54,[0,1,2,5,6,7,8,9,13,18],[1874.0,18.0,14.0,90.0,208.0,209.0,135.0,793.0,1.0,1.0])                 |
|(54,[0,1,2,3,4,5,6,7,8,9,13,18],[1879.0,28.0,19.0,30.0,12.0,95.0,209.0,196.0,117.0,778.0,1.0,1.0])   |
|(54,[0,1,2,3,4,5,6,7,8,9,13,15],[1888.0,33.0,22.0,150.0,46.0,108.0,209.0,185.0,103.0,735.0,1.0,1.0]) |
|(54,[0,1,2,3,4,5,6,7,8,9,13,14],[1889.0,28.0,22.0,150.0,23.0,120.0,205.0,185.0,108.0,759.0,1.0,1.0]) |
|(54,[0,1,2,3,4,5,6,7,8,9,13,18],[1896.0,337.0,12.0,30.0,6.0,175.0,195.0,224.0,168.0,732.0,1.0,1.0])  |
|(54,[0,1,2,3,4,5,6,7,8,9,13,15],[1898.0,34.0,23.0,175.0,56.0,13

In [23]:
assembledTrainData.getClass

res22: Class[T] = class org.apache.spark.sql.Dataset

In [24]:
trainData.getClass

res23: Class[T] = class org.apache.spark.sql.Dataset

Note on above. The transform fucntion converts a dataframe to a dataframe but re-organises it.

Now to create the DecisionTreem classifier for predicting the group cover type (a categorical valiable ranging 1..7)

In [25]:
import org.apache.spark.ml.classification.DecisionTreeClassifier
import scala.util.Random

import org.apache.spark.ml.classification.DecisionTreeClassifier

import scala.util.Random

In [26]:
val classifier = new org.apache.spark.ml.classification.DecisionTreeClassifier()
                        .setSeed(Random.nextLong())
                        .setLabelCol("Cover_Type")
                        .setFeaturesCol("featureVector")
                        .setPredictionCol("prediction")


classifier: DecisionTreeClassifier = dtc_637bd1ce331c

In [27]:
val model = classifier.fit(assembledTrainData)
//print(model.toDebugString)

take at Classifier.scala:111

4 / 4

take at Classifier.scala:111

1 / 1

take at DecisionTreeMetadata.scala:112

1 / 1

count at DecisionTreeMetadata.scala:118

4 / 4

flatMap at RandomForest.scala:919

4 / 4

collectAsMap at RandomForest.scala:922

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

model: org.apache.spark.ml.classification.DecisionTreeClassificationModel = DecisionTreeClassificationModel (uid=dtc_637bd1ce331c) of depth 5 with 49 nodes

In [28]:
print(model.toDebugString)

DecisionTreeClassificationModel (uid=dtc_637bd1ce331c) of depth 5 with 49 nodes
  If (feature 0 <= 3050.5)
   If (feature 0 <= 2552.5)
    If (feature 10 <= 0.5)
     If (feature 0 <= 2457.5)
      If (feature 3 <= 15.0)
       Predict: 4.0
      Else (feature 3 > 15.0)
       Predict: 3.0
     Else (feature 0 > 2457.5)
      If (feature 17 <= 0.5)
       Predict: 2.0
      Else (feature 17 > 0.5)
       Predict: 3.0
    Else (feature 10 > 0.5)
     If (feature 9 <= 5179.0)
      Predict: 2.0
     Else (feature 9 > 5179.0)
      If (feature 5 <= 559.0)
       Predict: 2.0
      Else (feature 5 > 559.0)
       Predict: 5.0
   Else (feature 0 > 2552.5)
    If (feature 0 <= 2956.5)
     If (feature 15 <= 0.5)
      If (feature 17 <= 0.5)
       Predict: 2.0
      Else (feature 17 > 0.5)
       Predict: 3.0
     Else (feature 15 > 0.5)
      Predict: 3.0
    Else (feature 0 > 2956.5)
     If (feature 3 <= 214.0)
      If (feature 36 <= 0.5)
       Predict: 2.0
      Else (feature 36 > 0.5)

Cool! Above cell appears to show a recursive definition (and almost self explaining) of the decision tree model.
So now the model has been fit (estimator(ed) in spark speak) we can use the model to make predictions on the test data. 
Note that the default settings have been used on the above model so we don't expect this out of the box model to do very well.

Without getting under the hood I assume that the decison tree optimises the gini index for the tree. TODO find out.

DecisiontTreeClassifiers have an 'importance' dunction to assess the importance of each feature. Not obvious what this is but assume its some sort of sensitivity analysis.??

In [29]:
model.featureImportances.toArray.zip(inputCols).sorted.reverse.foreach(println)

(0.7947580866849727,Elevation)
(0.04001267294058104,Horizontal_Distance_To_Hydrology)
(0.03136387156077785,Soil_Type_3)
(0.02794484080698811,Soil_Type_31)
(0.026601339258044594,Wilderness_Area_0)
(0.025800592086383592,Soil_Type_1)
(0.025743624221384813,Hillshade_Noon)
(0.01165509357972449,Wilderness_Area_2)
(0.006386560175755645,Horizontal_Distance_To_Roadways)
(0.006317752978144056,Soil_Type_22)
(0.002426984349214562,Hillshade_9am)
(9.885813580286153E-4,Horizontal_Distance_To_Fire_Points)
(0.0,Wilderness_Area_3)
(0.0,Wilderness_Area_1)
(0.0,Vertical_Distance_To_Hydrology)
(0.0,Soil_Type_9)
(0.0,Soil_Type_8)
(0.0,Soil_Type_7)
(0.0,Soil_Type_6)
(0.0,Soil_Type_5)
(0.0,Soil_Type_4)
(0.0,Soil_Type_39)
(0.0,Soil_Type_38)
(0.0,Soil_Type_37)
(0.0,Soil_Type_36)
(0.0,Soil_Type_35)
(0.0,Soil_Type_34)
(0.0,Soil_Type_33)
(0.0,Soil_Type_32)
(0.0,Soil_Type_30)
(0.0,Soil_Type_29)
(0.0,Soil_Type_28)
(0.0,Soil_Type_27)
(0.0,Soil_Type_26)
(0.0,Soil_Type_25)
(0.0,Soil_Type_24)
(0.0,Soil_Type_23)
(0.0,Soi

So the above model has the elevation as by far and away the most important feature. However we have not assessed the performance of modey yet so this should be treated cautiously until this is done. 

Note that the decisionTreeClassificationModel is a transformed ( in spark speak this is just a function that transform a dataframe into a dataframe

In [30]:
val preds = model.transform(assembledTrainData)

preds.select("Cover_Type","prediction","probability").show(truncate=false)

show at cmd29.sc:3

1 / 1

+----------+----------+------------------------------------------------------------------------------------------------+
|Cover_Type|prediction|probability                                                                                     |
+----------+----------+------------------------------------------------------------------------------------------------+
|6.0       |3.0       |[0.0,0.0,0.03696340723768448,0.6277714131680032,0.049599029584203785,0.0,0.2856661500101085,0.0]|
|6.0       |4.0       |[0.0,0.0,0.04659188955996549,0.27825711820534943,0.4301121656600518,0.0,0.2450388265746333,0.0] |
|6.0       |3.0       |[0.0,0.0,0.03696340723768448,0.6277714131680032,0.049599029584203785,0.0,0.2856661500101085,0.0]|
|6.0       |3.0       |[0.0,0.0,0.03696340723768448,0.6277714131680032,0.049599029584203785,0.0,0.2856661500101085,0.0]|
|6.0       |3.0       |[0.0,0.0,0.03696340723768448,0.6277714131680032,0.049599029584203785,0.0,0.2856661500101085,0.0]|
|6.0       |3.0       |[0.0,0.0,

preds: DataFrame = [Elevation: int, Aspect: int ... 57 more fields]

In [31]:
preds.columns.foreach(println)

Elevation
Aspect
Slope
Horizontal_Distance_To_Hydrology
Vertical_Distance_To_Hydrology
Horizontal_Distance_To_Roadways
Hillshade_9am
Hillshade_Noon
Hillshade_3pm
Horizontal_Distance_To_Fire_Points
Wilderness_Area_0
Wilderness_Area_1
Wilderness_Area_2
Wilderness_Area_3
Soil_Type_0
Soil_Type_1
Soil_Type_2
Soil_Type_3
Soil_Type_4
Soil_Type_5
Soil_Type_6
Soil_Type_7
Soil_Type_8
Soil_Type_9
Soil_Type_10
Soil_Type_11
Soil_Type_12
Soil_Type_13
Soil_Type_14
Soil_Type_15
Soil_Type_16
Soil_Type_17
Soil_Type_18
Soil_Type_19
Soil_Type_20
Soil_Type_21
Soil_Type_22
Soil_Type_23
Soil_Type_24
Soil_Type_25
Soil_Type_26
Soil_Type_27
Soil_Type_28
Soil_Type_29
Soil_Type_30
Soil_Type_31
Soil_Type_32
Soil_Type_33
Soil_Type_34
Soil_Type_35
Soil_Type_36
Soil_Type_37
Soil_Type_38
Soil_Type_39
Cover_Type
featureVector
rawPrediction
probability
prediction


In [32]:
val results = preds.select("Cover_Type","prediction")

results: DataFrame = [Cover_Type: double, prediction: double]

In [33]:
results.selectExpr("Cover_Type==prediction as Matched","Cover_Type")
    .groupBy("Cover_Type").pivot("Matched",Seq("true","false")).count().show()

show at cmd32.sc:2

4 / 4

show at cmd32.sc:2

200 / 200

show at cmd32.sc:2

1 / 1

show at cmd32.sc:2

4 / 4

show at cmd32.sc:2

20 / 20

show at cmd32.sc:2

100 / 100

show at cmd32.sc:2

75 / 75

+----------+------+-----+
|Cover_Type|  true|false|
+----------+------+-----+
|       7.0|  8567| 9831|
|       1.0|124421|65895|
|       4.0|   997| 1491|
|       3.0| 26125| 5960|
|       2.0|205939|48984|
|       6.0|  null|15581|
|       5.0|   128| 8390|
+----------+------+-----+



In [34]:
val predPivot = results.selectExpr("Cover_Type","prediction")
    .groupBy("Cover_Type")
    .pivot("prediction",(1 to 7))
    .count()
    .na.fill(0).sort("Cover_Type")

predPivot: Dataset[Row] = [Cover_Type: double, 1: bigint ... 6 more fields]

In [35]:
predPivot.show()

show at cmd34.sc:1

4 / 4

show at cmd34.sc:1

200 / 200

show at cmd34.sc:1

200 / 200

+----------+------+------+-----+---+---+---+----+
|Cover_Type|     1|     2|    3|  4|  5|  6|   7|
+----------+------+------+-----+---+---+---+----+
|       1.0|124421| 62183|  162|  0|  0|  0|3550|
|       2.0| 43706|205939| 4492|108| 65|  0| 613|
|       3.0|     0|  5315|26125|645|  0|  0|   0|
|       4.0|     0|    19| 1472|997|  0|  0|   0|
|       5.0|     4|  7630|  756|  0|128|  0|   0|
|       6.0|     0|  5829| 9184|568|  0|  0|   0|
|       7.0|  9576|   195|   60|  0|  0|  0|8567|
+----------+------+------+-----+---+---+---+----+



TODO: Add row total

In [36]:
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

In [37]:
val evaluator = new org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator()
                .setLabelCol("Cover_Type")
                .setPredictionCol("prediction")

evaluator: MulticlassClassificationEvaluator = mcEval_6d824aedfaaa

In [38]:
evaluator.setMetricName("accuracy").evaluate(preds)

map at MulticlassMetrics.scala:45

4 / 4

collectAsMap at MulticlassMetrics.scala:47

4 / 4

countByValue at MulticlassMetrics.scala:42

4 / 4

countByValue at MulticlassMetrics.scala:42

4 / 4

res37: Double = 0.7010735024669305

As expected accuracy not great. Also, looking at the prediction pivot table shows that there is class imbalance that might be affecting the utility of accuracy and f1 might be a better score.


In [39]:
evaluator.setMetricName("f1").evaluate(preds)

countByValue at MulticlassMetrics.scala:42

4 / 4

countByValue at MulticlassMetrics.scala:42

4 / 4

map at MulticlassMetrics.scala:45

4 / 4

collectAsMap at MulticlassMetrics.scala:47

4 / 4

map at MulticlassMetrics.scala:50

4 / 4

collectAsMap at MulticlassMetrics.scala:52

4 / 4

res38: Double = 0.6822695375053985

Actually not that difference suggesting that the algorithm (with the current set of parameters) isn't greatly affected by the class imbalance. This should be monitored on going as the imbalance of the data warrants this.

Its seems that spark has a confusion matrix ( what I produced in the above pivot table).Its accesses (in the book at least) via the Mllib rdd API (which is in maintenance mode now with the dataframe spark.ml api being the promoted one)

In [40]:
import org.apache.spark.mllib.evaluation.MulticlassMetrics
val predictionRDD = preds.
        select("prediction","Cover_Type").
        as[(Double,Double)].
        rdd
val multiclassMetrics = new org.apache.spark.mllib.evaluation.MulticlassMetrics(predictionRDD)

import org.apache.spark.mllib.evaluation.MulticlassMetrics

predictionRDD: org.apache.spark.rdd.RDD[(Double, Double)] = MapPartitionsRDD[132] at rdd at cmd39.sc:4
multiclassMetrics: MulticlassMetrics = org.apache.spark.mllib.evaluation.MulticlassMetrics@9d43a5e

In [41]:
multiclassMetrics.confusionMatrix

map at MulticlassMetrics.scala:45

4 / 4

collectAsMap at MulticlassMetrics.scala:47

4 / 4

map at MulticlassMetrics.scala:55

4 / 4

collectAsMap at MulticlassMetrics.scala:57

4 / 4

res40: org.apache.spark.mllib.linalg.Matrix = 124421.0  62183.0   162.0    0.0    0.0    0.0  3550.0  
43706.0   205939.0  4492.0   108.0  65.0   0.0  613.0   
0.0       5315.0    26125.0  645.0  0.0    0.0  0.0     
0.0       19.0      1472.0   997.0  0.0    0.0  0.0     
4.0       7630.0    756.0    0.0    128.0  0.0  0.0     
0.0       5829.0    9184.0   568.0  0.0    0.0  0.0     
9576.0    195.0     60.0     0.0    0.0    0.0  8567.0  

In [42]:
predPivot.show()

show at cmd41.sc:1

4 / 4

show at cmd41.sc:1

200 / 200

show at cmd41.sc:1

200 / 200

+----------+------+------+-----+---+---+---+----+
|Cover_Type|     1|     2|    3|  4|  5|  6|   7|
+----------+------+------+-----+---+---+---+----+
|       1.0|124421| 62183|  162|  0|  0|  0|3550|
|       2.0| 43706|205939| 4492|108| 65|  0| 613|
|       3.0|     0|  5315|26125|645|  0|  0|   0|
|       4.0|     0|    19| 1472|997|  0|  0|   0|
|       5.0|     4|  7630|  756|  0|128|  0|   0|
|       6.0|     0|  5829| 9184|568|  0|  0|   0|
|       7.0|  9576|   195|   60|  0|  0|  0|8567|
+----------+------+------+-----+---+---+---+----+



To evaluate the 70% accuracy from the default model we can benchmark it against a null model and use an inner product of the class probabilities to compare.

In [43]:
def classProbs(df: DataFrame):Array[Double] = {
    val totalCounts = df.count().asInstanceOf[Double]
        df.groupBy("Cover_Type").count()
        .orderBy("Cover_Type")
        .select("count").as[Double]
        .map(_/totalCounts)
        .collect()
}

defined function classProbs

Take inner product of test and train class probabilities as A measure of similarity

In [44]:
val trainProbs = classProbs(trainData)
val testProbs = classProbs(testData)
trainProbs.zip(testProbs).map(x=>x._1*x._2).sum

count at cmd42.sc:2

4 / 4

count at cmd42.sc:2

1 / 1

collect at cmd42.sc:6

4 / 4

collect at cmd42.sc:6

200 / 200

collect at cmd42.sc:6

200 / 200

collect at cmd42.sc:6

8 / 8

count at cmd42.sc:2

4 / 4

count at cmd42.sc:2

1 / 1

collect at cmd42.sc:6

4 / 4

collect at cmd42.sc:6

200 / 200

collect at cmd42.sc:6

200 / 200

collect at cmd42.sc:6

8 / 8

trainProbs: Array[Double] = Array(
  4.3441982944715773E-4,
  0.3642160540135034,
  0.48785729595664223,
  0.06140246796392976,
  0.004761394430240213,
  0.016301269194849734,
  0.029818041245005127,
  0.03520905736638241
)
testProbs: Array[Double] = Array(
  3.4058205473153617E-4,
  0.36653440730207926,
  0.48325187745857673,
  0.06247977794050032,
  0.004410537608773394,
  0.01660337516816239,
  0.03041397748752618,
  0.035965464979650225
)
res43_2: Double = 0.3755570819618255

The null model of randomly choosing a cover type label and using it as the prediction yields a 37% true positive rate. Makes the 70% TP rate look pretty good. 

## PIPELINES

In [45]:
import org.apache.spark.ml.Pipeline

import org.apache.spark.ml.Pipeline

In [55]:
val inputCols = trainData.columns.filter(_!="Cover_Type")
val assembler = new VectorAssembler().
                    setHandleInvalid("skip").
                    setInputCols(inputCols).
                    setOutputCol("featureVector")

inputCols: Array[String] = Array(
  "Elevation",
  "Aspect",
  "Slope",
  "Horizontal_Distance_To_Hydrology",
  "Vertical_Distance_To_Hydrology",
  "Horizontal_Distance_To_Roadways",
  "Hillshade_9am",
  "Hillshade_Noon",
  "Hillshade_3pm",
  "Horizontal_Distance_To_Fire_Points",
  "Wilderness_Area_0",
  "Wilderness_Area_1",
  "Wilderness_Area_2",
  "Wilderness_Area_3",
  "Soil_Type_0",
  "Soil_Type_1",
  "Soil_Type_2",
  "Soil_Type_3",
  "Soil_Type_4",
  "Soil_Type_5",
  "Soil_Type_6",
  "Soil_Type_7",
  "Soil_Type_8",
  "Soil_Type_9",
  "Soil_Type_10",
  "Soil_Type_11",
  "Soil_Type_12",
  "Soil_Type_13",
  "Soil_Type_14",
  "Soil_Type_15",
  "Soil_Type_16",
  "Soil_Type_17",
  "Soil_Type_18",
  "Soil_Type_19",
  "Soil_Type_20",
  "Soil_Type_21",
  "Soil_Type_22",
  "Soil_Type_23",
...
assembler: VectorAssembler = vecAssembler_da73fb4e8104

In [56]:
val classifer = new DecisionTreeClassifier().
                setSeed(Random.nextLong()).
                setLabelCol("Cover_Type").
                setFeaturesCol("featureVector").
                setPredictionCol("prediction")

classifer: DecisionTreeClassifier = dtc_97ecca588a8f

The above two transformations, recall that a transformation is a map from one dataframe to another that is part of the DAG, are chained together in a pipeline that is a convenience for defining a series of transforms on the data. We do this to streamline the hyper parameter tuning process.

In [57]:
val pipeline = new Pipeline().setStages(Array(assembler,classifier))

pipeline: Pipeline = pipeline_40fe55bb9817

Use spark.ml.tuning.ParameterBuilder to set up collections of hyper parameters to try. Will obviously need an evaluation metric as well. The ParamGrid BUilder defines the series of values to trye, and the loss function

In [58]:
import org.apache.spark.ml.tuning.{ParamGridBuilder,TrainValidationSplit}

import org.apache.spark.ml.tuning.{ParamGridBuilder,TrainValidationSplit}

In [59]:
val paramGrid = new ParamGridBuilder().
                    addGrid(classifer.impurity, Seq("gini","entropy")).
                    addGrid(classifier.maxDepth, Seq(1,20)).
                    addGrid(classifier.maxBins,Seq(40,300)).
                    addGrid(classifier.minInfoGain, Seq(0.0,0.05)).
                    build()



paramGrid: Array[org.apache.spark.ml.param.ParamMap] = Array(
  {
	dtc_97ecca588a8f-impurity: gini,
	dtc_637bd1ce331c-maxBins: 40,
	dtc_637bd1ce331c-maxDepth: 1,
	dtc_637bd1ce331c-minInfoGain: 0.0
},
  {
	dtc_97ecca588a8f-impurity: gini,
	dtc_637bd1ce331c-maxBins: 40,
	dtc_637bd1ce331c-maxDepth: 20,
	dtc_637bd1ce331c-minInfoGain: 0.0
},
  {
	dtc_97ecca588a8f-impurity: gini,
	dtc_637bd1ce331c-maxBins: 300,
	dtc_637bd1ce331c-maxDepth: 1,
	dtc_637bd1ce331c-minInfoGain: 0.0
},
  {
	dtc_97ecca588a8f-impurity: gini,
	dtc_637bd1ce331c-maxBins: 300,
	dtc_637bd1ce331c-maxDepth: 20,
	dtc_637bd1ce331c-minInfoGain: 0.0
},
  {
	dtc_97ecca588a8f-impurity: entropy,
	dtc_637bd1ce331c-maxBins: 40,
	dtc_637bd1ce331c-maxDepth: 1,
	dtc_637bd1ce331c-minInfoGain: 0.0
},
  {
	dtc_97ecca588a8f-impurity: entropy,
	dtc_637bd1ce331c-maxBins: 40,
	dtc_637bd1ce331c-maxDepth: 20,
	dtc_637bd1ce331c-minInfoGain: 0.0
},
  {
	dtc_97ecca588a8f-impurity: entropy,
...

In [60]:
val multiclassEval = new MulticlassClassificationEvaluator().
setLabelCol("Cover_Type").
setPredictionCol("prediction").
setMetricName("accuracy")

multiclassEval: MulticlassClassificationEvaluator = mcEval_ca793caf00a0

In [65]:
val validator = new TrainValidationSplit().
    setSeed(Random.nextLong()).
    setEstimator(pipeline).
    setEvaluator(multiclassEval).
    setEstimatorParamMaps(paramGrid).
    setTrainRatio(0.9)

validator: TrainValidationSplit = tvs_7a0e9879f8c6

Now go ahead and fit the model on all the parameters set in the setEstimatorParamMaps

In [ ]:
val validatorModel=validator.fit(trainData)

take at Classifier.scala:111

4 / 4

take at Classifier.scala:111

1 / 1

take at DecisionTreeMetadata.scala:112

1 / 1

count at DecisionTreeMetadata.scala:118

4 / 4

flatMap at RandomForest.scala:919

4 / 4

collectAsMap at RandomForest.scala:922

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

map at MulticlassMetrics.scala:45

4 / 4

collectAsMap at MulticlassMetrics.scala:47

4 / 4

countByValue at MulticlassMetrics.scala:42

4 / 4

countByValue at MulticlassMetrics.scala:42

4 / 4

take at Classifier.scala:111

4 / 4

take at Classifier.scala:111

1 / 1

take at DecisionTreeMetadata.scala:112

1 / 1

count at DecisionTreeMetadata.scala:118

4 / 4

flatMap at RandomForest.scala:919

4 / 4

collectAsMap at RandomForest.scala:922

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557

4 / 4

mapPartitions at RandomForest.scala:538

4 / 4

collectAsMap at RandomForest.scala:557 
 (kill)

1 + 3 / 4

The validator now contains the best model parameters for the pipeline supplied (in the setEstimator - recall that estimator is one of the spark.ml 'primatives' along with transform and is used for fitting parameters. Because a pipeline was supplied to the estimator. To get the model (parameters) out of the pipeline.

In [63]:
import org.apache.spark.ml.PipelineModel

import org.apache.spark.ml.PipelineModel

In [64]:
validatorModel.bestModel.asInstanceOf[PipelineModel].stages.last.extractParamMap

res63: org.apache.spark.ml.param.ParamMap = {
	dtc_637bd1ce331c-cacheNodeIds: false,
	dtc_637bd1ce331c-checkpointInterval: 10,
	dtc_637bd1ce331c-featuresCol: featureVector,
	dtc_637bd1ce331c-impurity: gini,
	dtc_637bd1ce331c-labelCol: Cover_Type,
	dtc_637bd1ce331c-maxBins: 300,
	dtc_637bd1ce331c-maxDepth: 20,
	dtc_637bd1ce331c-maxMemoryInMB: 256,
	dtc_637bd1ce331c-minInfoGain: 0.0,
	dtc_637bd1ce331c-minInstancesPerNode: 1,
	dtc_637bd1ce331c-predictionCol: prediction,
	dtc_637bd1ce331c-probabilityCol: probability,
	dtc_637bd1ce331c-rawPredictionCol: rawPrediction,
	dtc_637bd1ce331c-seed: 7387978820213730639
}

In [64]:
val best = validatorModel.bestModel

cmd64.sc:7: type arguments [_$7] do not conform to class Model's type parameter bounds [M <: org.apache.spark.ml.Model[M]]
  .printOnChange(best, "best", _root_.scala.None, _root_.scala.None, _root_.scala.None)) }
   ^cmd64.sc:7: type arguments [_$7] do not conform to class Model's type parameter bounds [M <: org.apache.spark.ml.Model[M]]
  .printOnChange(best, "best", _root_.scala.None, _root_.scala.None, _root_.scala.None)) }
                ^Compilation Failed

: 